dev-docs/services/ci/README.md

```diff
@@ -0,0 +1,244 @@
+    - Checks that the results of running the batch are what we would expect
+- The CI service polls the batch which it submits for overall success or failure.
+- Once the batch has either succeeded or failed, CI uses that result to report status back to GitHub
+
+Examples of CI test runs can be seen by searching through the production batch log, as long as you are a member
+of the `ci` billing project. For example: `/batches?q=user+%3D+ci%0D%0Atest+%3D+1`.
+
+#### CI Testing Timeline
+
+The image below shows the CI testing timeline:
+
+```mermaid
+sequenceDiagram
+    participant Github
+    box Kubernetes (PROD: 'default' namespace)
+    participant CI as PROD CI Service (default namespace)
+    participant PB as PROD Batch (default namespace)
+    end
+    box GCE (PROD VMs)
+    participant PVM as  PROD VMs
+    end
+    box Kubernetes (CI: 'test-xyz' namespaces)
+    participant CIB as CI Batch (test-xyz namespaces)
+    end
+    box GCE (CI VMs)
+    participant CIVM as CI VMs
+    end
+
+    Github->>CI: Notify PR created/updated
+
+    CI->>Github: Register github check (pending)
+    CI->>PB: Submit CI test batch
+    activate PB
+    PB->>PVM: Submit build jobs
+    activate PVM
+    PVM-->>PB: Done
+    deactivate PVM
+    PB->>PVM: Submit deploy jobs
+    activate PVM
+    PVM->>CIB: Deploy batch service
+    activate CIB
+    PVM-->>PB: Done
+    deactivate PVM
+    PB->>PVM: Submit test jobs
+    activate PVM
+    PVM->>CIB: Submit test batches
+    CIB->>CIVM: Submit test batch jobs
+    activate CIVM
+    CIVM-->>CIB: Validate results
+    deactivate CIVM
+    PVM-->>PB: Done
+    deactivate PVM
+    PB->>PVM: Submit cleanup jobs
+    activate PVM
+    PVM->>CIB: Clean up service
+    deactivate CIB
+    PVM-->>PB: Done
+    deactivate PVM
+    PB-->>CI: Detects completion
+    CI->>Github: Update github check
+```
+
+## Merging PRs to the `main` branch
+
+When a PR state changes, it will cause the `github_changed` flag to become dirty. 
+
+During its update loop, the CI service will determine whether any PRs targeting its watched branches are elibible to
+be merged.
+
+Readiness is determined by github status checks. The following conditions must be met:
+
+- The PR must be against the `main` branch
+- The PR must have passed all tests in GCP
+- The PR must be approved
+
+The control flow from final approval to CI merging a PRs looks like:
+
+- The PR's state will change in github (either a check changes to SUCCESS, or a review is approved)
+- The github webhook callback will cause the `github_changed` flag will be marked as dirty for the `WatchedBranch`
+- The `WatchedBranch`'s `_update` method in [`github.py`](../../../ci/ci/github.py) scans all PRs against the branch and updates state that helps determine mergeability.
+- The `WatchedBranch`'s `_update` method in [`github.py`](../../../ci/ci/github.py) iterates again to merge all mergeable PRs, in priority order
```

Comment:
There's a subtlety here which is being glossed over. As I understand (without having inspected the code, so correct me if I'm wrong), we keep a PR's check set to SUCCESS even if the successful CI run was on top of an outdated main (i.e. another PR has merged since this PR was last tested). Rerunning the tests of all PRs every time any PR merges would be wasteful, but we don't want to merge a PR unles it has a successful CI run on top of the current main. My mental model is that CI chooses one PR to be the next merge candidate, and reruns the tests if the last run is out of date, then merges only if that suceeds. But I don't know for sure if that's right.

Reply:
Hmm, interesting. I didn't see evidence for this in the code but it seems logical. 

I also wonder whether this is why we sometimes see PRs get "stuck" with pending checks. Perhaps CI is just choosing to delay re-running until the PR is back on top of its mergeable list 🤔

Reply:
I did a deep dive into this today, and you are indeed correct. I've added a flowchart showing the full set of paths through the update loop, which might be overkill in the end, but was a useful exercise for me in the process of making it.

What we seem to do is - 

- During `heal`:
  - Nominate a current merge candidate
    - If its tests are out of date, re-run them
- During `try_to_merge`:
  - Iterate over all PRs
    - Test is_mergeable 
    - Merge if mergeable (and return after the first success)

So... I think it's _possible_ that we might kick off a "merge candidate" test and then decide to merge some other PR while those tests are running (if its tests were already up to date)